### PR TITLE
GCC 11.3.1: use zstd for compressing LTO data

### PR DESCRIPTION
**Pull request by iarspider**
Notice: I have tried building gcc with static libzstd, but it didn't work (missing symbols at link-time)

### CODE REVIEW

**Comment by iarspider** [comment]
please test

**Comment by iarspider** [comment]
@smuzaffar can I tell cmsbot to use `-bootstrap` image for this? It needs cmake (to build zstd), but it's only in bootstrap image

**Comment by smuzaffar** [comment]
@iarspider , although I have added `cmake` to our container but why not just use `make`? `zstd` support it too ( https://github.com/facebook/zstd)

File: gcc.spec [hunk 1]
```diff
@@ -139,6 +142,44 @@ CFLAGS="${CONF_FLAGS}" ./configure --static --prefix=%{i}/tmp/sw
...
+ -DZSTD_BUILD_CONTRIB:BOOL=OFF \
+ -DZSTD_BUILD_STATIC:BOOL=OFF \
+ -DZSTD_BUILD_SHARED:BOOL=ON \
+ -DZSTD_BUILD_TESTS:BOOL=OFF \
+ -DCMAKE_BUILD_TYPE=Release \
+ -DZSTD_BUILD_PROGRAMS:BOOL=OFF \
+ -DZSTD_LEGACY_SUPPORT:BOOL=OFF \
+ -DCMAKE_INSTALL_PREFIX:STRING=%{i} \
+ -DCMAKE_INSTALL_LIBDIR:STRING=lib \
+ -Dzstd_VERSION:STRING=%{zstdVersion}
```
Comment (smuzaffar) [inline gcc.spec]: why this section of code? Can't you just copy files from `%{i}/tmp/sw` to `%{i}` ?

**Comment by iarspider** [comment]
I have just copied zstd recipe and mimicked [what we do for flex](https://github.com/cms-sw/cmsdist/blob/d7afb506700fcb3bd89b75250348c3b3bc22b49f/gcc.spec#L207-L222). I am not even sure we need to build zstd twice: maybe we can install it once to `%{i}` and just add `-I %{i}/include -L%{i}/lib` [here](https://github.com/cms-sw/cmsdist/blob/d7afb506700fcb3bd89b75250348c3b3bc22b49f/gcc.spec#L310)?

**Comment by smuzaffar** [comment]
what we do for flex is wrong, we should fix that too.  I think we can just use `make` and build static `zstd`.  I am looking at the zstd Makefile now and will provide a suggestion to use `make`

File: gcc.spec [hunk 2]
```diff
@@ -139,6 +142,44 @@ CFLAGS="${CONF_FLAGS}" ./configure --static --prefix=%{i}/tmp/sw
...
+
+make %{makeprocesses} VERBOSE=1
+make install
+
```
Comment (smuzaffar) [inline gcc.spec]:
```suggestion
#Build and install zstd static library
make -C ../zstd-%{zstdVersion}/lib %{makeprocesses} \
  install-static install-includes prefix=%{i}/tmp/sw \
  CPPFLAGS="-fPIC" CFLAGS="-fPIC"
```

try this locally and see if you can build gcc with zstd

**Comment by smuzaffar** [comment]
I have provided the change which should allow you to build and use zstd. While you are at it , please fix the flex build part too

File: gcc.spec [hunk 3]
```diff
@@ -263,13 +304,15 @@ export LD_LIBRARY_PATH=%{i}/lib64:%{i}/lib:$LD_LIBRARY_PATH
              $CONF_GCC_OS_SPEC $CONF_GCC_WITH_LTO --with-gmp=%{i} --with-mpfr=%{i} --enable-bootstrap \
              --with-mpc=%{i} --with-isl=%{i} --enable-checking=release \
              --build=%{_build} --host=%{_host} --enable-libstdcxx-time=rt $CONF_GCC_ARCH_SPEC \
-             --enable-shared --disable-libgcj CC="$CC" CXX="$CXX" CPP="$CPP" CXXCPP="$CXXCPP" \
+             --enable-shared --disable-libgcj \
+             --with-zstd-include=%{i}/tmp/sw/include --with-zstd-lib=%{i}/tmp/sw/lib \
```
Comment (smuzaffar) [inline gcc.spec]:
```suggestion
             --with-zstd=%{i}/tmp/sw \
```

did you try this?

**Comment by iarspider** [comment]
please test

**Comment by iarspider** [comment]
@smuzaffar looks like we either need to link statically to flex, or put it under `%{i}`

**Comment by smuzaffar** [comment]
@iarspider , I think adding `--disable-shared --enable-static` to `flex` configure command should be enough.

**Comment by iarspider** [comment]
please test

**Comment by smuzaffar** [comment]
hold
lets include these changes when we update gcc to 11.3 or above

**Comment by smuzaffar** [comment]
please test

**Comment by smuzaffar** [comment]
please test

**Comment by smuzaffar** [comment]
closing in favor of #8545 (which contains these changes along with gcc 11.4 update)

**Pull request closed without merging**